File: 323-Number-of-Connected-Components-in-an-Undirected-Graph/UnionFind.py
class UnionFind:
    def __init__(self, num: int) -> None:
        self.parent = list(range(num))
        self.size = [1] * num

    def root(self, i: int) -> int:
        if i != self.parent[i]:
            self.parent[i] = self.root(self.parent[i])
        return self.parent[i]

    def merge(self, i: int, j: int) -> None:
        i, j = self.root(i), self.root(j)
        if i == j:
            return
        if self.size[i] < self.size[j]:
            i, j = i, j
        self.parent[j] = i
        self.size[i] += self.size[j]


class Solution:
    def countComponents(self, n: int, edges: List[List[int]]) -> int:
        unionFind = UnionFind(n)
        for a, b in edges:
            unionFind.merge(a, b)
        return sum(unionFind.parent[i] == i for i in range(n))

File: medium/261-Graph-Valid-Tree/UnionFind.java
class Solution {
    public boolean validTree(int n, int[][] edges) {
        if (edges.length != n - 1) {
            return false;
        }
        
        UnionFind unionFind = new UnionFind(n);
        
        for (int[] edge: edges) {
            if (!unionFind.merge(edge[0], edge[1])) {
                return false;
            }
        }
        
        return true;
        
    }
}

class UnionFind {
    private int[] parent;
    private int[] size;
    
    public UnionFind(int num) {
        parent = new int[num];
        size = new int[num];
        for (int i = 0; i < num; i++) {
            parent[i] = i;
            size[i] = 1;
        }
    }
    
    private int root(int node) {
        if (node != parent[node]) {
            parent[node] = root(parent[node]);
        }
        return parent[node];
    }
    
    public boolean merge(int i, int j) {
        int iRoot = root(i);
        int jRoot = root(j);
        
        if (iRoot == jRoot) {
            return false;
        }
        
        if (parent[iRoot] < parent[jRoot]) {
            int tmp = jRoot;
            jRoot = iRoot;
            iRoot = tmp;
        }
        
        parent[jRoot] = iRoot;
        size[iRoot] += size[jRoot];
        
        return true;
    }
}
